internal/middleware/requestid.go

```diff
@@ -0,0 +1,56 @@
+package middleware
+
+import (
+	"context"
+
+	"github.com/google/uuid"
+	grpc_middleware "github.com/grpc-ecosystem/go-grpc-middleware"
+	"github.com/openfga/openfga/pkg/logger"
+	"go.uber.org/zap"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/metadata"
+)
+
+const requestIDCtxKey ctxKey = "request-id-context-key"
+
+func NewRequestIDInterceptor(logger logger.Logger) grpc.UnaryServerInterceptor {
+	return func(ctx context.Context, req interface{}, info *grpc.UnaryServerInfo, handler grpc.UnaryHandler) (interface{}, error) {
+		id, err := uuid.NewRandom()
+		if err != nil {
+			logger.Error("failed to generate uuid", zap.Error(err))
+		}
+
+		requestID := id.String()
+
+		ctx = context.WithValue(ctx, requestIDCtxKey, requestID)
+
+		err = grpc.SetHeader(ctx, metadata.Pairs("X-Request-Id", requestID))
+		if err != nil {
+			logger.Error("failed to set header", zap.Error(err))
+		}
+
+		return handler(ctx, req)
+	}
+}
+
+func NewStreamingRequestIDInterceptor(logger logger.Logger) grpc.StreamServerInterceptor {
```

Comment:
Mind adding a simple test for this? Just to make sure that the extracted context includes the request id.

Reply:
Yep. Done. thanks!